Separate function bodies from their signatures in HIR

flodiebold opened the pull request:
Also give them their own dep map node.

I'm still unhappy with the handling of inlined items (1452edc1), but maybe you have a suggestion how to improve it.

Fixes #35078.

r? @nikomatsakis

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @nikomatsakis (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

nikomatsakis commented on src/librustc_incremental/calculate_svh/mod.rs:
```diff
@@ -149,19 +149,27 @@ impl<'a, 'tcx> HashItemsVisitor<'a, 'tcx> {
         assert!(def_id.is_local());
         debug!("HashItemsVisitor::calculate(def_id={:?})", def_id);
+        self.calculate_def_hash(DepNode::Hir(def_id), false, &mut walk_op);
+        self.calculate_def_hash(DepNode::HirBody(def_id), true, &mut walk_op);
```
If we were using `DepNode::Hir` for everything, then this code would not (I think) have to change; we would probably just never walk bodies in the hasher. Instead, we would call `calculate_def_id` on the bodies of fns as we traversed.

nikomatsakis commented on src/test/incremental/hello_world.rs:
```diff
@@ -31,7 +31,7 @@ mod x {
     use x;
 
-    #[rustc_dirty(label="TypeckItemBody", cfg="rpass2")]
+    #[rustc_clean(label="TypeckItemBody", cfg="rpass2")]
```
Nice =)

nikomatsakis commented on src/test/incremental/source_loc_macros.rs:
```diff
@@ -17,17 +17,17 @@
 }
 
-#[rustc_clean(label="Hir", cfg="rpass2")]
+#[rustc_clean(label="HirBody", cfg="rpass2")]
```
Seems like it'd be good to have both `Hir` and `HirBody` here?

nikomatsakis commented on src/librustc/middle/cstore.rs:
```diff
@@ -144,6 +144,7 @@ impl<'a> InlinedItemRef<'a> {
     pub fn from_trait_item(def_id: DefId, item: &'a hir::TraitItem, _map: &hir_map::Map) -> InlinedItemRef<'a> {
         let (body, kind) = match item.node {
             hir::ConstTraitItem(ref ty, Some(ref body)) => (&**body, InlinedItemKindRef::Const(ty)),
+            hir::ConstTraitItem(_, None) => bug!("InlinedItemRef::from_trait_item called for const without body"),
```
Is this a pre-existing bug, or something you introduced?

flodiebold replied:
The code which calls `from_trait_item` takes care to only call it when there's a body, but I accidentally changed that once, and it wasn't immediately clear to me what I'd done wrong, so I just added this special message in case it happens again.

nikomatsakis commented on src/librustc_typeck/collect.rs:
```diff
@@ -128,28 +128,32 @@ struct CollectItemTypesVisitor<'a, 'tcx: 'a> {
-    fn visit_item(&mut self, item: &hir::Item) {
+impl<'a, 'tcx> Visitor<'tcx> for CollectItemTypesVisitor<'a, 'tcx> {
+    fn nested_visit_map(&mut self) -> Option<(&hir::map::Map<'tcx>, NestedVisitMode)> {
+        Some((&self.ccx.tcx.map, NestedVisitMode::OnlyBodies))
```
Hmm, this is suboptimal from the POV of the dep-graph tasks. We don't want the task which writes to the signature of a fn also reading from the fn body. Not sure what's the best way to tweak that just now; it's certainly possible.

nikomatsakis commented on src/test/incremental/hello_world.rs:
```diff
@@ -18,22 +18,23 @@ fn main() { }
-    #[rustc_clean(label="TypeckItemBody", cfg="rpass2")]
-    pub fn y() {
-        x::x();
+    // FIXME: This should be clean
```
I think the problem is the issue of collect reading from the fn body that I pointed out earlier.

eddyb commented on src/librustc/middle/cstore.rs:
```diff
@@ -100,18 +100,100 @@ pub struct NativeLibrary {
-    Item(DefId /* def-id in source crate */, P<hir::Item>),
-    TraitItem(DefId /* impl id */, P<hir::TraitItem>),
-    ImplItem(DefId /* impl id */, P<hir::ImplItem>)
+pub struct InlinedItem {
```
Seems okay - shouldn't need the type of a constant though, in fact, you should be able to only record a `Vec<DefId>` for the arguments of a `const fn`, which doesn't even need to be in a separate enum anymore, just replace the `kind` field with `const_fn_args: Vec<DefId>` and that should be that.

flodiebold replied:
That would make implementing `FnLike` even harder, though, I think.

flodiebold replied:
`lookup_const_fn_by_id` returns an `FnLikeNode` currently. But I guess it could return something else, e.g. a new enum?

eddyb replied:
Yeah, that would make more sense, `FnLikeNode` is more about pretty-printing.

flodiebold commented:
Giving def-ids to bodies didn't really occur to me (I was thinking only definitions can have def-ids). That would probably make a few things cleaner, but I don't really have the perspective to know what else it may complicate. Also, how would the testing attributes work? Can you give attributes to expressions?

flodiebold commented on src/librustc_passes/consts.rs:
```diff
@@ -181,11 +181,19 @@ impl<'a, 'gcx> CheckCrateVisitor<'a, 'gcx> {
-                                      fn_like.body(),
-                                      fn_like.span(),
-                                      fn_like.id());
+            let qualif = match self.tcx.const_qualif_map.borrow_mut().entry(fn_like.body().node_id()) {
```
Ah, yeah. Because the `FnLike` interface is only half implemented for inlined items, the `fn_like` call would fail for them, but the `const_qualif_map` is already populated for them, so I added the check beforehand. I'll need to clean that up somehow...

nikomatsakis commented:
@flodiebold so I pushed two commits into your branch -- take a look. The first one fixes the `TypeckCollect` dependency problem in a reasonable clean way. The other one is a WIP -- basically a bunch of the "svh" tests are now much cleaner (in particular, a lot of once-dirty metadata and signatures are now clean) and I didn't get time to make them all pass.

nikomatsakis commented:
> Can you give attributes to expressions?

You can give attributes to statements, not expressions. So that is a possible complication; I guess we could look for an attribute on the first statement of a block or something.

nikomatsakis commented:
@flodiebold all in all, the main thing I'd prefer to settle now is the visitor trait -- i.e., whether to have another method. The `DepNode::HirBody` thing is relatively easy to refactor later, I think.

bors commented:
:umbrella: The latest upstream changes (presumably #37642) made this pull request unmergeable. Please resolve the merge conflicts.

flodiebold commented:
@nikomatsakis It does seem to me like having two methods and having no default for `nested_visit_map` would be cleaner. I'll do that.

I'll fix the remaining tests; I think they may have come in during my last rebase, after which I forgot to check them again :)

nikomatsakis commented:
@flodiebold 

> It does seem to me like having two methods and having no default for nested_visit_map would be cleaner. I'll do that.

Great =)

If that change is made and tests are working, I think this PR is ready to go. We can revisit the `HirBody` thing afterwards. I tend to think it'd be mildly cleaner to just have one variant (`Hir`) but I think the diff will be non-trivial-enough we might as well handle it in a follow-up (we have to create def-ids for fn bodies, at minimum? I feel like when I thought it through I decided there would be some (minor, but still) complications to overcome.)

flodiebold commented:
@eddyb I refactored the inlined item handling again. I didn't get rid of the `FnLikeNode` usage for local const fns though, just for the inlined items.

@nikomatsakis Ok, I think I've addressed everything.

flodiebold commented:
@michaelwoerister By the way, the `unary_and_binary_exprs` test was actually not working correctly because `cfail2` was misspelled `cfails2` in the `rustc_clean` annotations. Maybe that could be prevented somehow?

michaelwoerister commented:
@flodiebold Good catch. Yes, maybe we can let the rest runner do some checks to prevent this.

bors commented:
:umbrella: The latest upstream changes (presumably #37676) made this pull request unmergeable. Please resolve the merge conflicts.

eddyb commented:
@flodiebold The commit "Merge master into separate-bodies" is worrying me - did you `git pull` without `--rebase`? In other news, I'm blocked on this PR, more or less, so let me know if I can help.

flodiebold commented:
@eddyb I did merge because I wasn't sure whether a merge or a rebase is the preferred way to resolve conflicts. But I've actually just done a rebase onto the current master, so I could push that too.

Apart from that, I think the PR is ready and just needs to be looked at again.

eddyb commented:
@flodiebold Rebase + force push is the way, only @bors does merges (into the `auto` branch).
I've also used merge *locally* to temporarily create a branch made from multiple PRs to test something.

flodiebold commented:
Ok, that's what I preferred anyway. Rebased version coming up.

nikomatsakis commented:
@flodiebold I was experimenting with an alternative version of the visitor refactoring. I've just rebased it atop your rebase, once it builds I'll push it to your branch and you can tell me what you think. It's actually sort of closer to your original approach.

nikomatsakis commented:
done.

nikomatsakis commented:
Pushed one more commit updating the comments a bit.

nikomatsakis commented:
@flodiebold (if you are happy with my latest commit, then r=me on the PR)

flodiebold commented:
Ah, that does seem nicer. Makes it easier to search for visitors which currently don't return a map, too. :+1:

nikomatsakis commented:
@bors r+

bors commented:
:pushpin: Commit f620072 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis f620072891bbc7c77362ea71332f8d5f38c79dc8 -->

eddyb commented:
@bors r- [failed Travis](https://travis-ci.org/rust-lang/rust/jobs/179551926#L1126) (`librustdoc` changes needed after rebase)

bors commented:
:umbrella: The latest upstream changes (presumably #37791) made this pull request unmergeable. Please resolve the merge conflicts.

flodiebold commented:
@eddyb @nikomatsakis Fixed and rebased :)

michaelwoerister commented:
@bors r=nikomatsakis

bors commented:
:pushpin: Commit 593b273 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 593b2736598f3e08bb7636615e52e1f76a2f2da5 -->

bors commented:
:hourglass: Testing commit 593b273 with merge f50dbd5...

nikomatsakis commented:
@flodiebold great, thanks!

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](https://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/11145), [auto-linux-32-opt](https://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/11207), [auto-linux-64-cargotest](https://buildbot.rust-lang.org/builders/auto-linux-64-cargotest/builds/2432), [auto-linux-64-cross-freebsd](https://buildbot.rust-lang.org/builders/auto-linux-64-cross-freebsd/builds/2252), [auto-linux-64-debug-opt](https://buildbot.rust-lang.org/builders/auto-linux-64-debug-opt/builds/4419), [auto-linux-64-nopt-t](https://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/11149), [auto-linux-64-opt](https://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/11098), [auto-linux-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-linux-64-opt-rustbuild/builds/3077), [auto-linux-64-x-android-t](https://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/11072), [auto-linux-cross-opt](https://buildbot.rust-lang.org/builders/auto-linux-cross-opt/builds/4298), [auto-linux-musl-64-opt](https://buildbot.rust-lang.org/builders/auto-linux-musl-64-opt/builds/6250), [auto-mac-32-opt](https://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/11191), [auto-mac-64-opt](https://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/11164), [auto-mac-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-mac-64-opt-rustbuild/builds/3110), [auto-mac-cross-ios-opt](https://buildbot.rust-lang.org/builders/auto-mac-cross-ios-opt/builds/2287), [auto-win-gnu-32-opt](https://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/6293), [auto-win-gnu-32-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt-rustbuild/builds/3144), [auto-win-gnu-64-opt](https://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/6256), [auto-win-msvc-32-opt](https://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/5442), [auto-win-msvc-64-cargotest](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-cargotest/builds/2454), [auto-win-msvc-64-opt](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/6267), [auto-win-msvc-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-rustbuild/builds/3133)
Approved by: nikomatsakis
Pushing f50dbd580f3d5d88a300da540db3c85164bfd0de to master...
